In [221]:
import pandas as pd
import json

# 📌 Extracción

Para iniciar mi análisis, importaré los datos de la API de Telecom X. Estos datos están disponibles en formato JSON y contienen información esencial sobre los clientes, incluyendo datos demográficos, tipo de servicio contratado y estado de evasión.

✅ Cargando los datos directamente desde la API utilizando Python.

✅ Normalizando las columnas con diccionarios.

✅ Convirtiendo los datos a un DataFrame de Pandas para facilitar su manipulación.

In [222]:
with open('../data/TelecomX_Data.json','r') as f:
    data = json.loads(f.read())

In [223]:
pd_data = pd.DataFrame(data)

# 🔧 Transformación

## Conociendo el conjunto de datos

Ahora que extraí los datos, es fundamental comprender la estructura del dataset y el significado de sus columnas. Esta etapa ayudará a identificar qué variables son más relevantes para el análisis de evasión de clientes.

📌 Para facilitar este proceso, hay en el README.md un diccionario de datos con la descripción de cada columna.

¿Qué debo hacer?

✅ Explorar las columnas del dataset y verificar sus tipos de datos.

✅ Consultar el diccionario para comprender mejor el significado de las variables.

✅ Identificar las columnas más relevantes para el análisis de evasión.



In [224]:
pd_data.sample(3)

customerID Churn                                           customer  \
6036  8237-ULIXL    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
3097  4315-MURBD   Yes  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1703  2418-TPEUN    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   

                                                phone  \
6036  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
3097   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1703   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                               internet  \
6036  {'InternetService': 'No', 'OnlineSecurity': 'N...   
3097  {'InternetService': 'Fiber optic', 'OnlineSecu...   
1703  {'InternetService': 'No', 'OnlineSecurity': 'N...   

                                                account  
6036  {'Contract': 'Two year', 'PaperlessBilling': '...  
3097  {'Contract': 'Month-to-month', 'PaperlessBilli...  
1703  {'Contract': 'Two year', 'PaperlessBilling': '...

## Normalizando nuestros datos

Vemos que las columnas poseen diccionarios, por lo que normalizaré para separarlos en nuevas columnas, almacenandolo en un nuevo DataFrame para poder analizar y comparar los datos.

In [225]:
data_norm = pd.json_normalize(data)

In [226]:
data_norm.sample(3)

customerID Churn customer.gender  customer.SeniorCitizen  \
3170  4425-OWHWB    No          Female                       0   
5680  7762-URZQH   Yes            Male                       0   
2048  2884-GBPFB   Yes          Female                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
3170               No                  No               20                Yes   
5680              Yes                  No               66                Yes   
2048              Yes                  No               35                Yes   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
3170                  No                       No  ...   No internet service   
5680                 Yes              Fiber optic  ...                    No   
2048                  No              Fiber optic  ...                    No   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
3170       No internet service  No internet service  No internet service   
5680                       Yes                   No                  Yes   
2048                        No                   No                   No   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
3170      No internet service         One year                      Yes   
5680                      Yes         Two year                      Yes   
2048                       No   Month-to-month                      Yes   

        account.PaymentMethod account.Charges.Monthly  account.Charges.Total  
3170  Credit card (automatic)                   19.35                 433.75  
5680  Credit card (automatic)                  106.05                6981.35  
2048         Electronic check                   70.30                2416.55  

[3 rows x 21 columns]

Vemos los tipos de datos almacenados en las columnas

In [227]:
data_norm.dtypes

customerID                    object
Churn                         object
customer.gender               object
customer.SeniorCitizen         int64
customer.Partner              object
customer.Dependents           object
customer.tenure                int64
phone.PhoneService            object
phone.MultipleLines           object
internet.InternetService      object
internet.OnlineSecurity       object
internet.OnlineBackup         object
internet.DeviceProtection     object
internet.TechSupport          object
internet.StreamingTV          object
internet.StreamingMovies      object
account.Contract              object
account.PaperlessBilling      object
account.PaymentMethod         object
account.Charges.Monthly      float64
account.Charges.Total         object
dtype: object

Observamos que la columna **account.Charges.Total** tiene datos de tipo *object*, entonces la convertimos a tipo *float64*.

In [228]:
data_norm['account.Charges.Total'] = pd.to_numeric(data_norm['account.Charges.Total'], errors='coerce')

In [229]:
data_norm.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7267 non-null   object 
 1   Churn                      7267 non-null   object 
 2   customer.gender            7267 non-null   object 
 3   customer.SeniorCitizen     7267 non-null   int64  
 4   customer.Partner           7267 non-null   object 
 5   customer.Dependents        7267 non-null   object 
 6   customer.tenure            7267 non-null   int64  
 7   phone.PhoneService         7267 non-null   object 
 8   phone.MultipleLines        7267 non-null   object 
 9   internet.InternetService   7267 non-null   object 
 10  internet.OnlineSecurity    7267 non-null   object 
 11  internet.OnlineBackup      7267 non-null   object 
 12  internet.DeviceProtection  7267 non-null   object 
 13  internet.TechSupport       7267 non-null   objec

Observamos que nuestro df es de 7267 filas por 21 columnas.

In [230]:
data_norm.shape

(7267, 21)

En total tenemos 152607 datos

In [231]:
data_norm.size

152607

Vemos una lista de las columnas.

In [232]:
data_norm.columns

Index(['customerID', 'Churn', 'customer.gender', 'customer.SeniorCitizen',
       'customer.Partner', 'customer.Dependents', 'customer.tenure',
       'phone.PhoneService', 'phone.MultipleLines', 'internet.InternetService',
       'internet.OnlineSecurity', 'internet.OnlineBackup',
       'internet.DeviceProtection', 'internet.TechSupport',
       'internet.StreamingTV', 'internet.StreamingMovies', 'account.Contract',
       'account.PaperlessBilling', 'account.PaymentMethod',
       'account.Charges.Monthly', 'account.Charges.Total'],
      dtype='object')

## Comprobación de incoherencias en los datos

En este paso, verifico si hay problemas en los datos que puedan afectar el análisis. Prestando atención a valores ausentes, duplicados, errores de formato e inconsistencias en las categorías. Este proceso es esencial para asegurarme de que los datos estén listos para las siguientes etapas.

In [233]:
data_norm.describe()

customer.SeniorCitizen  customer.tenure  account.Charges.Monthly  \
count             7267.000000      7267.000000              7267.000000   
mean                 0.162653        32.346498                64.720098   
std                  0.369074        24.571773                30.129572   
min                  0.000000         0.000000                18.250000   
25%                  0.000000         9.000000                35.425000   
50%                  0.000000        29.000000                70.300000   
75%                  0.000000        55.000000                89.875000   
max                  1.000000        72.000000               118.750000   

       account.Charges.Total  
count            7256.000000  
mean             2280.634213  
std              2268.632997  
min                18.800000  
25%               400.225000  
50%              1391.000000  
75%              3785.300000  
max              8684.800000

Vemos que 'Churn' tiene 3 valores posibles, cuando sólo deberían ser 2.

In [234]:
data_norm.nunique()

customerID                   7267
Churn                           3
customer.gender                 2
customer.SeniorCitizen          2
customer.Partner                2
customer.Dependents             2
customer.tenure                73
phone.PhoneService              2
phone.MultipleLines             3
internet.InternetService        4
internet.OnlineSecurity         3
internet.OnlineBackup           3
internet.DeviceProtection       3
internet.TechSupport            3
internet.StreamingTV            3
internet.StreamingMovies        3
account.Contract                3
account.PaperlessBilling        2
account.PaymentMethod           4
account.Charges.Monthly      1585
account.Charges.Total        6530
dtype: int64

Vemos los valores únicos por columna para corroborar valores

In [235]:
for columna in data_norm.columns:
    print(f'{columna} \t {data_norm[columna].unique()}')

customerID 	 ['0002-ORFBO' '0003-MKNFE' '0004-TLHLJ' ... '9992-UJOEL' '9993-LHIEB'
 '9995-HOTOH']
Churn 	 ['No' 'Yes' '']
customer.gender 	 ['Female' 'Male']
customer.SeniorCitizen 	 [0 1]
customer.Partner 	 ['Yes' 'No']
customer.Dependents 	 ['Yes' 'No']
customer.tenure 	 [ 9  4 13  3 71 63  7 65 54 72  5 56 34  1 45 50 23 55 26 69 11 37 49 66
 67 20 43 59 12 27  2 25 29 14 35 64 39 40  6 30 70 57 58 16 32 33 10 21
 61 15 44 22 24 19 47 62 46 52  8 60 48 28 41 53 68 51 31 36 17 18 38 42
  0]
phone.PhoneService 	 ['Yes' 'No']
phone.MultipleLines 	 ['No' 'Yes' 'No phone service']
internet.InternetService 	 ['DSL' 'Fiber optic' 'No' 'Fibjsoner optic']
internet.OnlineSecurity 	 ['No' 'Yes' 'No internet service']
internet.OnlineBackup 	 ['Yes' 'No' 'No internet service']
internet.DeviceProtection 	 ['No' 'Yes' 'No internet service']
internet.TechSupport 	 ['Yes' 'No' 'No internet service']
internet.StreamingTV 	 ['Yes' 'No' 'No internet service']
internet.StreamingMovies 	 ['No' 'Yes' 'No 

⚠️ En 'Churn' no debería haber valores vacíos.

Verificamos duplicados de filas

In [236]:
sum(int(x) for x in data_norm.duplicated().values)

0

No hay filas duplicadas

Verificamos si hay un sólo registro para cada cliente

In [237]:
len(data_norm['customerID'].unique()) == len(data_norm)

True

No hay más de un registro para un cliente

## Manejo de Inconsistencias

Aplico las correcciones necesarias. Ajusto los datos para asegurarme de que estén completos y coherentes, preparándolos para las siguientes etapas del análisis.

In [238]:
data_norm.sample(3)

customerID Churn customer.gender  customer.SeniorCitizen  \
7054  9715-SBVSU    No            Male                       0   
1803  2569-WGERO    No          Female                       0   
2924  4065-JJAVA    No          Female                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
7054              Yes                 Yes               14                Yes   
1803               No                  No               72                Yes   
2924               No                  No                1                Yes   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
7054                  No                      DSL  ...                   Yes   
1803                  No                       No  ...   No internet service   
2924                  No                      DSL  ...                   Yes   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
7054                        No                   No                   No   
1803       No internet service  No internet service  No internet service   
2924                        No                   No                   No   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
7054                      Yes         Two year                      Yes   
1803      No internet service         Two year                      Yes   
2924                       No   Month-to-month                       No   

          account.PaymentMethod account.Charges.Monthly  account.Charges.Total  
7054  Bank transfer (automatic)                   61.40                 815.55  
1803  Bank transfer (automatic)                   21.15                1419.40  
2924           Electronic check                   49.50                  49.50  

[3 rows x 21 columns]

Buscamos la fila con 'Churn' == ''

In [239]:
data_norm[data_norm['Churn']=='']

customerID Churn customer.gender  customer.SeniorCitizen  \
30    0047-ZHDTW                Female                       0   
75    0120-YZLQA                  Male                       0   
96    0154-QYHJU                  Male                       0   
98    0162-RZGMZ                Female                       1   
175   0274-VVQOQ                  Male                       1   
...          ...   ...             ...                     ...   
7158  9840-GSRFX                Female                       0   
7180  9872-RZQQB                Female                       0   
7211  9920-GNDMB                  Male                       0   
7239  9955-RVWSC                Female                       0   
7247  9966-VYRTZ                Female                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
30                 No                  No               11                Yes   
75                 No                  No               71                Yes   
96                 No                  No               29                Yes   
98                 No                  No                5                Yes   
175               Yes                  No               65                Yes   
...               ...                 ...              ...                ...   
7158               No                  No               14                Yes   
7180              Yes                  No               49                 No   
7211               No                  No                9                Yes   
7239              Yes                 Yes               67                Yes   
7247              Yes                 Yes               31                Yes   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
30                   Yes              Fiber optic  ...                    No   
75                    No                       No  ...   No internet service   
96                    No                      DSL  ...                   Yes   
98                    No                      DSL  ...                   Yes   
175                  Yes              Fiber optic  ...                   Yes   
...                  ...                      ...  ...                   ...   
7158                 Yes                      DSL  ...                   Yes   
7180    No phone service                      DSL  ...                    No   
7211                 Yes              Fiber optic  ...                    No   
7239                  No                       No  ...   No internet service   
7247                  No                       No  ...   No internet service   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
30                          No                   No                   No   
75         No internet service  No internet service  No internet service   
96                          No                  Yes                   No   
98                          No                  Yes                   No   
175                        Yes                   No                  Yes   
...                        ...                  ...                  ...   
7158                        No                   No                   No   
7180                        No                   No                  Yes   
7211                        No                   No                   No   
7239       No internet service  No internet service  No internet service   
7247       No internet service  No internet service  No internet service   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
30                         No   Month-to-month                      Yes   
75        No internet service         Two year                      Yes   
96                         No         One year                      Yes   
98                         No   Month-to-month                    

Eliminaremos esos 224 registros

Obtenemos una lista con sus índices

In [240]:
filas_a_eliminar = data_norm[data_norm['Churn']==''].index

In [241]:
data_norm.drop(filas_a_eliminar, inplace=True)
data_norm.reset_index(drop=True, inplace=True)

Verifico si no quedaron registros con 'Churn' == '', y si la cantidad restante es correcta

In [242]:
( len(data_norm[data_norm['Churn']==''].index) == 0 ) & ( len(data_norm) == ( len(pd_data) - len(filas_a_eliminar) ) )

True

In [243]:
data_norm

customerID Churn customer.gender  customer.SeniorCitizen  \
0     0002-ORFBO    No          Female                       0   
1     0003-MKNFE    No            Male                       0   
2     0004-TLHLJ   Yes            Male                       0   
3     0011-IGKFF   Yes            Male                       1   
4     0013-EXCHZ   Yes          Female                       1   
...          ...   ...             ...                     ...   
7038  9987-LUTYD    No          Female                       0   
7039  9992-RRAMN   Yes            Male                       0   
7040  9992-UJOEL    No            Male                       0   
7041  9993-LHIEB    No            Male                       0   
7042  9995-HOTOH    No            Male                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
0                 Yes                 Yes                9                Yes   
1                  No                  No                9                Yes   
2                  No                  No                4                Yes   
3                 Yes                  No               13                Yes   
4                 Yes                  No                3                Yes   
...               ...                 ...              ...                ...   
7038               No                  No               13                Yes   
7039              Yes                  No               22                Yes   
7040               No                  No                2                Yes   
7041              Yes                 Yes               67                Yes   
7042              Yes                 Yes               63                 No   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
0                     No                      DSL  ...                   Yes   
1                    Yes                      DSL  ...                    No   
2                     No              Fiber optic  ...                    No   
3                     No              Fiber optic  ...                   Yes   
4                     No              Fiber optic  ...                    No   
...                  ...                      ...  ...                   ...   
7038                  No                      DSL  ...                    No   
7039                 Yes              Fiber optic  ...                    No   
7040                  No                      DSL  ...                   Yes   
7041                  No                      DSL  ...                    No   
7042    No phone service                      DSL  ...                   Yes   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
0                           No                  Yes                  Yes   
1                           No                   No                   No   
2                          Yes                   No                   No   
3                          Yes                   No                  Yes   
4                           No                  Yes                  Yes   
...                        ...                  ...                  ...   
7038                        No                  Yes                   No   
7039                        No                   No                   No   
7040                        No                   No                   No   
7041                       Yes                  Yes                   No   
7042                       Yes                   No                  Yes   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
0                          No         One year                      Yes   
1                         Yes   Month-to-month                       No   
2                          No   Month-to-month                      Yes   
3                         Yes   Month-to-month                    

## Columna de cuentas diarias

Ahora que los datos están limpios, creamos la columna "Cuentas_Diarias" solicitada por el desafío, bajo el nombre "account.Charges.Daily". Usamos la facturación mensual para calcular el valor diario, proporcionando una visión más detallada del comportamiento de los clientes a lo largo del tiempo.

In [249]:
data_norm['account.Charges.Daily'] = (data_norm['account.Charges.Monthly'] / 30).round(2)
data_norm.head(3)

customerID Churn customer.gender  customer.SeniorCitizen customer.Partner  \
0  0002-ORFBO    No          Female                       0              Yes   
1  0003-MKNFE    No            Male                       0               No   
2  0004-TLHLJ   Yes            Male                       0               No   

  customer.Dependents  customer.tenure phone.PhoneService phone.MultipleLines  \
0                 Yes                9                Yes                  No   
1                  No                9                Yes                 Yes   
2                  No                4                Yes                  No   

  internet.InternetService  ... internet.DeviceProtection  \
0                      DSL  ...                        No   
1                      DSL  ...                        No   
2              Fiber optic  ...                       Yes   

  internet.TechSupport internet.StreamingTV internet.StreamingMovies  \
0                  Yes                  Yes                       No   
1                   No                   No                      Yes   
2                   No                   No                       No   

  account.Contract account.PaperlessBilling account.PaymentMethod  \
0         One year                      Yes          Mailed check   
1   Month-to-month                       No          Mailed check   
2   Month-to-month                      Yes      Electronic check   

  account.Charges.Monthly account.Charges.Total  account.Charges.Daily  
0                    65.6                593.30                   2.19  
1                    59.9                542.40                   2.00  
2                    73.9                280.85                   2.46  

[3 rows x 22 columns]

## Estandarización y Transformación de Datos

La estandarización y transformación de datos es una etapa opcional, pero altamente recomendada, ya que busca hacer que la información sea más consistente, comprensible y adecuada para el análisis. Durante esta fase, por ejemplo, podemos convertir valores textuales como "Sí" y "No" en valores binarios (1 y 0), lo que facilita el procesamiento matemático y la aplicación de modelos analíticos.

Además, traducir o renombrar columnas y datos hace que la información sea más accesible y fácil de entender, especialmente cuando se trabaja con fuentes externas o términos técnicos. Aunque no es un paso obligatorio, puede mejorar significativamente la claridad y comunicación de los resultados, facilitando la interpretación y evitando confusiones, especialmente al compartir información con stakeholders no técnicos.

Los nombres de las columnas decido dejarlos en inglés, me resulta más compacto a la vista, que su correspondiente al español.

Pero sí modificaré las columnas:
```
{
    'Churn': {'No': 0, 'Yes': 1},
    'customer.gender': {'Male': 0, 'Female': 1},
    'customer.Partner': {'No': 0, 'Yes': 1},
    'customer.Dependents': {'No': 0, 'Yes': 1},
    'phone.PhoneService': {'No': 0, 'Yes': 1},
    'phone.MultipleLines': {'No': 0, 'Yes': 1, 'No phone service': 2},
    'internet.InternetService': {'No': 0, 'DSL': 1, 'Fiber optic': 2, 'Fibjsoner optic': 3},
    'internet.OnlineSecurity': {'No': 0, 'Yes': 1, 'No internet service': 2},
    'internet.OnlineBackup': {'No': 0, 'Yes': 1, 'No internet service': 2},
    'internet.DeviceProtection': {'No': 0, 'Yes': 1, 'No internet service': 2},
    'internet.TechSupport': : {'No': 0, 'Yes': 1, 'No internet service': 2},
    'internet.StreamingTV': {'No': 0, 'Yes': 1, 'No internet service': 2},
    'internet.StreamingMovies': {'No': 0, 'Yes': 1, 'No internet service': 2},
    'account.Contract': {'One year': 12, 'Month-to-month': 1, 'Two year': 24},
    'account.PaperlessBilling':	{'No': 0, 'Yes': 1},
    'account.PaymentMethod': {'Mailed check': 0, 'Electronic check': 1, 'Credit card (automatic)': 2, 'Bank transfer (automatic)': 3}
}
```


In [258]:
diccionario_reemplazo_multiple = {
    'Churn': {'No': 0, 'Yes': 1},
    'customer.gender': {'Male': 0, 'Female': 1},
    'customer.Partner': {'No': 0, 'Yes': 1},
    'customer.Dependents': {'No': 0, 'Yes': 1},
    'phone.PhoneService': {'No': 0, 'Yes': 1},
    'phone.MultipleLines': {'No': 0, 'Yes': 1, 'No phone service': 2},
    'internet.InternetService': {'No': 0, 'DSL': 1, 'Fiber optic': 2, 'Fibjsoner optic': 3},
    'internet.OnlineSecurity': {'No': 0, 'Yes': 1, 'No internet service': 2},
    'internet.OnlineBackup': {'No': 0, 'Yes': 1, 'No internet service': 2},
    'internet.DeviceProtection': {'No': 0, 'Yes': 1, 'No internet service': 2},
    'internet.TechSupport': {'No': 0, 'Yes': 1, 'No internet service': 2},
    'internet.StreamingTV': {'No': 0, 'Yes': 1, 'No internet service': 2},
    'internet.StreamingMovies': {'No': 0, 'Yes': 1, 'No internet service': 2},
    'account.Contract': {'One year': 12, 'Month-to-month': 1, 'Two year': 24},
    'account.PaperlessBilling':	{'No': 0, 'Yes': 1},
    'account.PaymentMethod': {'Mailed check': 0, 'Electronic check': 1, 'Credit card (automatic)': 2, 'Bank transfer (automatic)': 3}
}


In [259]:
data_norm = data_norm.replace(diccionario_reemplazo_multiple)
data_norm.sample(3)

C:\Users\Usuario\AppData\Local\Temp\ipykernel_11312\1891672484.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  data_norm = data_norm.replace(diccionario_reemplazo_multiple)


customerID  Churn  customer.gender  customer.SeniorCitizen  \
5339  7542-CYDDM      0                0                       0   
436   0637-YLETY      1                1                       0   
2623  3733-LSYCE      0                1                       0   

      customer.Partner  customer.Dependents  customer.tenure  \
5339                 0                    0               18   
436                  0                    0               16   
2623                 1                    0               15   

      phone.PhoneService  phone.MultipleLines  internet.InternetService  ...  \
5339                   1                    0                         0  ...   
436                    1                    1                         2  ...   
2623                   1                    0                         2  ...   

      internet.DeviceProtection  internet.TechSupport  internet.StreamingTV  \
5339                          2                     2                     2   
436                           1                     0                     0   
2623                          0                     0                     0   

      internet.StreamingMovies  account.Contract  account.PaperlessBilling  \
5339                         2                 1                         0   
436                          1                 1                         1   
2623                         0                 1                         0   

      account.PaymentMethod  account.Charges.Monthly  account.Charges.Total  \
5339                      0                    20.05                 358.50   
436                       2                    95.60                1555.65   
2623                      3                    75.35                1114.55   

      account.Charges.Daily  
5339                   0.67  
436                    3.19  
2623                   2.51  

[3 rows x 22 columns]

#📊 Carga y análisis

#📄Informe final